[#164829362] Social login implementation backend facing

kirega opened the pull request:
#### Description
Adds user authentication with social login for Google, Facebook, and Twitter. 
Users who wish to use any one of these social platforms with automatically get access to the site after successfully being authorized by these 3rd party sites. 
#### Type of change
Separates the logic for handling the way that the different social platforms have structured their social logins.  There are three separate endpoints for each social site that only accept a `POST`  method of the following nature:
##### `POST` `localhost/api/users/google/`
- Login using google 
```
{
         "access_token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IjZmNjc4MWJhNzExOTlhNjU4ZTc2MGFhNWFhOTNlNWZjM2RjNzUyYjUiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJhY2NvdW50cy5nb29nbGUuY29tIiwiYXpwIjoiNDk5MTIxNjI4Mzc2LXU2bjQ1OW4ybGVkNTJzMHE0MWg4aTc1MXFzNjk2N29zLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwiYXVkIjoiNDk5MTIxNjI4Mzc2LXU2bjQ1OW4ybGVkNTJzMHE0MWg4aTc1MXFzNjk2N29zLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwic3ViIjoiMTE1MzAzNTgxMzk0NTEyNjgyNjAwIiwiZW1haWwiOiJqb2VraXJlZ2FAZ21haWwuY29tIiwiZW1haWxfdmVyaWZpZWQiOnRydWUsImF0X2hhc2giOiJRZDJVUmZZZGdValRqTXZmQXFoZXdnIiwibmFtZSI6ImpvZSBraXJlZ2EiLCJwaWN0dXJlIjoiaHR0cHM6Ly9saDUuZ29vZ2xldXNlcmNvbnRlbnQuY29tLy1rcTBGMFZCSm9Tcy9BQUFBQUFBQUFBSS9BQUFBQUFBQUFBQS9BQ0hpM3JlYTFaNm50LXFaTUVqVk5JeW1yRmlBQXdVeVFnL3M5Ni1jL3Bob3RvLmpwZyIsImdpdmVuX25hbWUiOiJqb2UiLCJmYW1pbHlfbmFtZSI6ImtpcmVnYSIsImxvY2FsZSI6ImVuIiwiaWF0IjoxNTU0ODQyOTM3LCJleHAiOjE1NTQ4NDY1MzcsImp0aSI6ImI1MzYxMGEyOTFkNmYzNWYyMzhlYjRjZDNjNzdmNjEzZWMwNjMyOTIifQ.r5yXGPAL04uwsNczquypReGLyP6saHpgXdxNZT2kBUe5zKWditS25VKfXJjmCK6vnz1oJm8W5WItEDq8M0wOK_0gM9i1UMXjzKpCVB4RAYVvSoUu6nKWeV7TUu0BBSFY-MnyM62VoHekzQdNnmqW14lunX6dP6iyMsM6MEV51MYi53VwM96odrZ1e2HVWn_5dwWvef4HkkP8knrsbRp1YFp15odB37RMWQzl2n1WF4Id5BubrlMiA7LvsbdthwPnBhIs11PcOabEqZmSxjk4N0Zss2yEofKsBpiuri5umfBHXrjDwtq9EFuuZWNz6veqjImMYLbPG5B1f-SRyH647Q"
} 
```
On success the payload should return a user access token from out backend in the form of :
```

{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```

- Login using facebook 
##### `POST` `localhost/api/users/facebook/`
```
{
         "access_token": "< access_token>"
} 
```
On success the payload should return a user access token from out backend in the form of :
```
{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```
- Login using Twitter 
##### `POST` `localhost/api/users/twitter/`
This endpoint utilizes a slightly different work flow because it uses Oauth1. 
To make the `POST` send the `access_token_key` and `access_token_secret` separated with a space  a single payload
```
{
         "access_token": "< access_token_key> <access_token_secret >"
} 
```
On success the payload should return a user access token from out backend in the form of :
```
{
    "user": {
        "username": "<token>",
        "token": "<token>"
    }
}
```
Incase of invalid tokens the endpoint returns the error in the form of 
```
{
    "errors": {
        "access_token": [
            "Invalid token please try again"
        ]
    }
}
```
#### How Can This Be Tested?
- Clone the repo and navigate to the root directory
- Create a virtual environment by typing `virtualenv venv` and activate it
- Install requirements by typing `pip install -r dev-requirements.txt`
- Setup a database and ensure that your `.env` has the following variables:
```
export SECRET_KEY
export DEBUG
export DATABASE_URL
export SOCIAL_AUTH_GOOGLE_OAUTH2_KEY
export SOCIAL_AUTH_GOOGLE_OAUTH2_SECRET
export SOCIAL_AUTH_TWITTER_KEY
export SOCIAL_AUTH_TWITTER_SECRET
export AH_HOST
export CLOUDINARY_NAME
export CLOUDINARY_API_KEY
export CLOUDINARY_API_SECRET
export SENDGRID_API_KEY
export FROM_EMAIL
export URL
export EMAIL_HOST=
export EMAIL_HOST_PASSWORD
export EMAIL_PORT
export EMAIL_USE_TLS
export DOMAIN

```
- Run the tests by typing `python manage.py test`
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164829370

kirega committed:
[ft #164829362] Create Social_id field in models

kirega committed:
[ft #164829362] Create Social_id field in models

kirega committed:
[ft #164829362] Create Social_id field in models

kirega committed:
[ft #164829362] unistall social django

kirega committed:
[ft #164829362] unistall social django

kirega committed:
[ft #164829362] unistall social django

kirega committed:
[ft #164829362] unistall social django

kirega committed:
[ft #164829362] Add views for social login

kirega committed:
[ft #164829362] Add views for social login

kirega committed:
[ft #164829362] Add views for social login

kirega committed:
[ft #164829362] Add views for social login

kirega committed:
Merge branch 'develop' into ft-social-login-#164829362

kirega committed:
[ft ##164829362] Refactor the views and serializers

kirega committed:
[ft ##164829362] Refactor the views and serializers

kirega committed:
[ft ##164829362] Update social serializer

Adds more functionality in the way that serializer function

Kirega

kirega committed:
[ft ##164829362] Update social serializer

Adds more functionality in the way that serializer function

Kirega

kirega committed:
[ft ##164829362] Remove redundant selenium test

Following refactors to social login selenium rests are redundant

Kirega

kirega committed:
[ft ##164829362] Remove redundant selenium test

Following refactors to social login selenium rests are redundant

Kirega

kirega committed:
[ft ##164829362] Write tests for social login

Updates the body of tests needed for the social login

Kirega

kirega committed:
[ft ##164829362] Write tests for social login

Updates the body of tests needed for the social login

Kirega

kirega committed:
Merge branch 'ft-social-login-#164829362' of https://github.com/andela/ah-the-answer-backend into ft-social-login-#164829362

# Conflicts:
#	authors/apps/authentication/serializers.py
#	authors/apps/authentication/urls.py
#	authors/apps/authentication/validators.py
#	authors/apps/authentication/views.py
#	authors/settings.py
#	requirements.txt

kirega committed:
Merge branch 'ft-social-login-#164829362' of https://github.com/andela/ah-the-answer-backend into ft-social-login-#164829362

# Conflicts:
#	authors/apps/authentication/serializers.py
#	authors/apps/authentication/urls.py
#	authors/apps/authentication/validators.py
#	authors/apps/authentication/views.py
#	authors/settings.py
#	requirements.txt

kirega committed:
Merge branch 'ft-social-login-#164829362' of https://github.com/andela/ah-the-answer-backend into ft-social-login-#164829362

# Conflicts:
#	authors/apps/authentication/serializers.py
#	authors/apps/authentication/urls.py
#	authors/apps/authentication/validators.py
#	authors/apps/authentication/views.py
#	authors/settings.py
#	requirements.txt

Raywire reviewed (commented):
I have tested the google login locally it is working, please update the payload to include more details like username or email the user on login.
<img width="1131" alt="Screenshot 2019-04-10 at 12 31 13" src="https://user-images.githubusercontent.com/8179091/55868226-b31bbf80-5b8c-11e9-94ff-3fb11801a9d9.png">

kirega commented:
@Raywire  I appreciate that you took time to review. I considered deeply the pros and cons of having the details of the user alongside the token and I came to the conclusion that it was unnecessary to add the user detail. The reasoning behind this conclusion is that the token is already encoded with user details. I hope this makes sense and that you may consider my approach.

kirega commented:
@Raywire I appreciate that you took time to review. As I consider to add in more user details. Would you mind elaborating on the benefit of taking such a course of action?

kihuha reviewed (dismissed):
Functionality works

WNjihia committed:
Merge branch 'develop' into ft-social-login-#164829362

WNjihia commented:
Kindly update the travis.yml to remove the selenium setup

kirega committed:
[#164829362] Refactor: Update twitter

Refactors the twitter feature to use requests-oauthlib

[#164829370] Pivotal Tracker

kirega committed:
[#164829362] Update travis.yml

Removes the selenium and chrome drivers

[#164829370] Pivotal Tracker

kirega committed:
Merge branch 'develop' into ft-social-login-#164829362

kirega committed:
[#164829362] Update tests

Updates breaking tests following refactors

[#164829370] Pivotal Tracker

kirega committed:
[#164829362] Update tests

Updates breaking tests following refactors

[#164829370] Pivotal Tracker

kirega committed:
Merge remote-tracking branch 'origin/ft-social-login-#164829362' into ft-social-login-#164829362

kihuha reviewed (dismissed):
Works okay

Raywire reviewed (dismissed):
Social auth works well and returns a valid token that can be used to log in to the app.

WNjihia committed:
Merge branch 'develop' into ft-social-login-#164829362

kihuha commented:
Quick one, I am trying to access the profiles endpoints and they require a username to be appended to the url. How would I achieve this if I have logged in using any social auth platform? Is there a way of getting the username from the social auth token?

WNjihia commented on requirements.txt:
```diff
@@ -61,3 +61,7 @@ soupsieve==1.9
 typed-ast==1.3.1
 urllib3==1.24.1
 wrapt==1.11.1
+selenium==3.141.0
```
Do we still need to have selenium as a requirement?

kirega replied:
No we do not use this anymore. I will remove it.

kirega committed:
[finishes #164829362] Update requirements.txt

Remove selenium from dependencies

[#164829370] Pivotal Tracker

kirega commented:
@kihuha Nice point you have raised there. I had not envisioned that scenario. I can nest the users details alongside the login token to assist you with that.

kirega committed:
[finishes #164829362] Include username in login response in social

Includes the username in the response while logging in.

[#164829362] Pivotal Tracker

kirega commented:
I believe with this new build the username will be incorporated

kirega commented:
@kihuha You will find the issue fixed in the latest commit

WNjihia committed:
Merge branch 'develop' into ft-social-login-#164829362

kirega committed:
Merge branch 'develop' into ft-social-login-#164829362

kirega committed:
[#164829362] Increase app coverage

Kyppy reviewed (dismissed):
Good job correcting the PR instructions.

Raywire reviewed (dismissed):
Social auth works well and returns a valid token that can be used to log in to the app.

kirega committed:
[#164829370] Fix a bug

Raywire reviewed (approved):
I am satisfied by how this feature works in the backend, tested it locally everything is in order, great job.

ipaullly reviewed (approved):
have run the functionality locally, its an interesting feature and works as you described it.
